api/v1/submissions/PKPSubmissionFileHandler.inc.php

```diff
@@ -0,0 +1,455 @@
+		$fileId = Services::get('file')->add(
+			$_FILES['file']['tmp_name'],
+			$submissionDir . '/' . uniqid() . '.' . $extension
+		);
+
+		$params = $this->convertStringsToSchema(SCHEMA_SUBMISSION_FILE, $slimRequest->getParsedBody());
+		$params['fileId'] = $fileId;
+		$params['submissionId'] = $submission->getId();
+		$params['uploaderUserId'] = (int) $request->getUser()->getId();
+
+		$primaryLocale = $request->getContext()->getPrimaryLocale();
+		$allowedLocales = $request->getContext()->getData('supportedSubmissionLocales');
+
+		// Set the name if not passed with the request
+		if (empty($params['name'])) {
+			$params['name'][$primaryLocale] = $_FILES['file']['name'];
+		}
+
+		// If no genre has been set and there is only one genre possible, set it automatically
+		if (empty($params['genreId'])) {
+			$genres = DAORegistry::getDAO('GenreDAO')->getEnabledByContextId($request->getContext()->getId());
+			if ($genres->count === 1) {
+				$params['genreId'] = $genres->next()->getId();
+			}
+		}
+
+		$errors = Services::get('submissionFile')->validate(VALIDATE_ACTION_ADD, $params, $allowedLocales, $primaryLocale);
+
+		if (!empty($errors)) {
+			return $response->withStatus(400)->withJson($errors);
+		}
+
+		// Review attachments and discussion files can not be uploaded through this API endpoint
+		$notAllowedFileStages = [
+			SUBMISSION_FILE_NOTE,
+			SUBMISSION_FILE_REVIEW_ATTACHMENT,
+			SUBMISSION_FILE_QUERY,
+		];
+		if (in_array($params['fileStage'], $notAllowedFileStages)) {
+			return $response->withStatus(400)->withJsonError('api.submissionFiles.403.unauthorizedFileStageIdWrite');
+		}
+
+		// A valid review round is required when uploading to a review file stage
+		$reviewFileStages = [
+			SUBMISSION_FILE_INTERNAL_REVIEW_FILE,
+			SUBMISSION_FILE_INTERNAL_REVIEW_REVISION,
+			SUBMISSION_FILE_REVIEW_FILE,
+			SUBMISSION_FILE_REVIEW_REVISION,
+		];
+		if (in_array($params['fileStage'], $reviewFileStages)) {
+			if (empty($params['assocType']) || $params['assocType'] !== ASSOC_TYPE_REVIEW_ROUND || empty($params['assocId'])) {
+				return $response->withStatus(400)->withJsonError('api.submissionFiles.400.missingReviewRoundAssocType');
+			}
+			$reviewRoundDao = DAORegistry::getDAO('ReviewRoundDAO'); /* @var $reviewRoundDao ReviewRoundDAO */
+			$reviewRound = $reviewRoundDao->getById($params['assocId']);
+			$stageId = in_array($params['fileStage'], [SUBMISSION_FILE_INTERNAL_REVIEW_FILE, SUBMISSION_FILE_INTERNAL_REVIEW_REVISION])
+				? WORKFLOW_STAGE_ID_INTERNAL_REVIEW
+				: WORKFLOW_STAGE_ID_EXTERNAL_REVIEW;
+			if (!$reviewRound
+					|| $reviewRound->getData('submissionId') != $params['submissionId']
+					|| $reviewRound->getData('stageId') != $stageId) {
+				return $response->withStatus(400)->withJsonError('api.submissionFiles.400.reviewRoundSubmissionNotMatch');
+			}
+		}
+
+		$submissionFile = DAORegistry::getDao('SubmissionFileDAO')->newDataObject();
+		$submissionFile->_data = $params;
```

Comment:
Use the `DataObject::setAllData` function rather than accessing `_data` directly.

Reply:
(Beware, `setAllData` takes a reference. At a glance that should not be necessary, but if you opt to remove the reference from the function, please make sure to adjust the tests and review current use cases.)

Reply:
Didn't know that existed! I've opted to remove the reference and will keep an eye for the tests.